torchvision/prototype/datasets/utils/_resource.py

```diff
@@ -57,7 +54,42 @@ def _extract(file: pathlib.Path) -> None:
     def _decompress(file: pathlib.Path) -> None:
         _decompress(str(file), remove_finished=True)
 
-    def _loader(self, path: pathlib.Path) -> IterDataPipe[Tuple[str, IO]]:
+    @classmethod
+    def from_http(cls, url: str, *, file_name: Optional[str] = None, **kwargs: Any) -> OnlineResource:
+        return cls(url, file_name=file_name or pathlib.Path(urlparse(url).path).name, **kwargs)
+
+    @classmethod
+    def from_gdrive(cls, id: str, **kwargs: Any) -> OnlineResource:
+        return cls(f"https://drive.google.com/uc?export=download&id={id}", **kwargs)
+
+    def download(self, root: Union[str, pathlib.Path], *, skip_integrity_check: bool = False) -> pathlib.Path:
+        root = pathlib.Path(root).expanduser().resolve()
+        root.mkdir(parents=True, exist_ok=True)
+        file = root / self.file_name
+
+        if not file.exists():
+            dp = IterableWrapper([self.url])
+            dp = OnlineReader(dp)
+            stream = list(dp)[0][1]
+
+            with open(file, "wb") as fh, tqdm() as progress_bar:
+                for chunk in iter(lambda: stream.read(1024 * 1024), b""):  # type: ignore[no-any-return]
+                    # filter out keep-alive new chunks
+                    if not chunk:
+                        continue
+
+                    fh.write(chunk)
+                    progress_bar.update(len(chunk))
+
+        if self.sha256 and not skip_integrity_check:
+            dp = IterableWrapper([str(file)])
+            dp = FileOpener(dp, mode="rb")
+            dp = HashChecker(dp, {str(file): self.sha256}, hash_type="sha256")
+            list(dp)
```

Comment:
Even if we could do the above in datapipes, we cannot attach the `HashChecker` directly. It will consume the stream and if it is not seekable, e.g. a HTTP response, there is no data left to save afterwards. This can probably be done with a `OnDiskCacheHolder`, but let's not worry about that if we are already breaking out of the graph above.

Reply:
I was able to solve this with a `Forker` in 2d03f026d2c04c38b23540f95c71f2006e4fbbcb.